// styled-spinkit/dist/index.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0});var e,t=require("react"),a=require("styled-components"),o=(e=a)&&"object"==typeof e&&"default"in e?e.default:e;const r=(e,t)=>Math.round(e*10**t)/10**t,s=(e,t=e)=>a.css`
  width: ${e};
  height: ${t};
`,i=e=>s(`${e}px`),n=({size:e})=>i(e),l=({color:e})=>a.css`
  background-color: ${e};
`,c=({size:e=0})=>a.css`
  margin: ${e}px auto;
`,m=e=>a.css`
    animation-delay: ${e}s;
  `,d=({delay:e})=>m(e),$=e=>Array.from(new Array(e).keys()),p=a.keyframes`
  100% {
    transform: rotate(360deg);
  }
`,f=a.keyframes`
  0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
`,u=o.div`
  ${s("60%")};
  display: inline-block;
  position: absolute;
  top: 0;
  border-radius: 100%;
  animation: ${f} ${2}s infinite ease-in-out;
  ${e=>e.second&&a.css`
      top: auto;
      bottom: 0;
      ${m(-1)};
    `};
`,y=o.div`
  ${n};
  ${c};
  position: relative;
  text-align: center;
  animation: ${p} ${2}s infinite linear;

  > ${u} {
    ${l};
  }
`,g=a.keyframes`
  0%,
  39%,
  100% {
    opacity: 0;
  }

  40% {
    opacity: 1;
  }
`,x=o.div`
  ${s("100%")};
  position: absolute;
  left: 0;
  top: 0;
  transform: rotate(${e=>e.rotate}deg);
  &::before {
    content: '';
    display: block;
    margin: 0 auto;
    ${s("15%")};
    border-radius: 100%;
    animation: ${g} 1.2s infinite ease-in-out both -${e=>e.delay}s;
  }
`,b=o.div`
  ${n};
  ${c};
  position: relative;

  > ${x} {
    &::before {
      ${l};
    }
  }
`,v=a.keyframes`
  0%,
  80%,
  100% {
    transform: scale(0);
  }
  40% {
    transform: scale(1);
  }
`,z=o.div`
  border-radius: 100%;
  display: inline-block;
  animation: ${v} 1.4s ease-in-out ${e=>e.delay}s infinite both;
`,k=o.div`
  width: ${e=>e.size}px;
  ${c};
  text-align: center;
  border-radius: 100%;
  > ${z} {
    ${e=>i(e.size/4)};
    ${l};
  }
`,E=a.keyframes`
  0%,
  80%,
  100% {
    transform: scale(0);
  }
  40% {
    transform: scale(1);
  }
`,N=o.div`
  ${s("100%")};
  position: absolute;
  top: 0;
  left: 0;
  transform: rotate(${e=>e.transform}deg);

  &::before {
    content: '';
    display: block;
    margin: 0 auto;
    ${s("15%")};
    border-radius: 100%;
    animation: ${E} 1.2s ease-in-out ${e=>e.delay}s infinite both;
  }
`,h=o.div`
  ${c};
  ${n};
  position: relative;

  > ${N} {
    &::before {
      ${l};
    }
  }
`,X=a.keyframes`
  0%,
  10% {
    transform: perspective(140px) rotateX(-180deg);
    opacity: 0;
  }
  25%,
  75% {
    transform: perspective(140px) rotateX(0deg);
    opacity: 1;
  }
  90%,
  100% {
    transform: perspective(140px) rotateY(180deg);
    opacity: 0;
  }
`,Y=o.div`
  float: left;
  ${s("50%")};
  position: relative;
  transform: scale(1.1) rotateZ(${e=>e.scale}deg);
  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    ${s("100%")};
    animation: ${X} ${e=>e.duration}s infinite linear both
      ${e=>e.delay}s;
    transform-origin: 100% 100%;
  }
`,C=o.div`
  ${n};
  ${c};
  position: relative;
  transform: rotateZ(45deg);

  > ${Y} {
    &::before {
      ${l};
    }
  }
`,D=o.div`
  ${n};
  ${d};
  position: absolute;
  top: 0;
  left: 0;
  animation-timing-function: ease-in-out;
  animation-iteration-count: infinite;
  animation-fill-mode: both;
`,F=o.div`
  margin: ${e=>r(.8*e.size,1)}px auto;
  ${n};
  text-align: center;
  font-size: 10px;
  position: relative;

  > ${D} {
    ${l};
    animation-name: ${e=>((e=42)=>a.keyframes`
  0% {
    transform: rotate(0deg);
  }
  25% {
    transform: translateX(${e}px) rotate(-90deg) scale(0.5);
  }
  50% {
    /* Hack to make FF rotate in the right direction */
    transform: translateX(${e}px) translateY(${e}px) rotate(-179deg);
  }
  50.1% {
    transform: translateX(${e}px) translateY(${e}px) rotate(-180deg);
  }
  75% {
    transform: translateX(0) translateY(${e}px) rotate(-270deg) scale(0.5);
  }
  100% {
    transform: rotate(-360deg);
  }
`)(r(.8*e.size,1))};
    animation-duration: ${e=>e.speed}s;
  }
`,w=a.keyframes`
  0%,
  70%,
  100% {
    transform: scale3d(1, 1, 1);
  }

  35% {
    transform: scale3d(0, 0, 1);
  }
`,P=o.div`
  ${s("33.33%")};
  float: left;
  animation: ${w} 1.3s infinite ease-in-out ${e=>e.delay}s;
`,j=o.div`
  ${n};
  ${c};
  > ${P} {
    ${l};
  }
`,q=a.keyframes`
  0% {
    transform: scale(0);
  }
  100% {
    opacity: 0;
    transform: scale(1);
  }
`,A=o.div`
  ${n};
  ${l};
  ${c};
  border-radius: 100%;
  animation: ${q} 1s ease-in-out infinite;
`,B=a.keyframes`
  0%,
  40%,
  100% {
    transform: scaleY(0.4);
  }
  20% {
    transform: scaleY(1);
  }
`,M=o.div`
  ${s("7px","100%")};
  margin: 0 3px 0 0;
  display: inline-block;
  ${d};
`,W=o.div`
  ${c};
  ${e=>s(`${1.25*e.size}px`,`${e.size}px`)};
  text-align: center;
  font-size: 10px;

  > ${M} {
    ${l};
    animation-name: ${B};
    animation-duration: ${e=>e.speed}s;
    animation-timing-function: ease-in-out;
    animation-iteration-count: infinite;
  }
`,Z=a.keyframes`
  0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
`,_=o.div`
  ${s("100%")};
  border-radius: 50%;
  position: absolute;
  opacity: 0.6;
  top: 0;
  left: 0;
  animation: ${Z} 2s infinite ease-in-out;
  ${e=>e.isDelay&&m(-1)};
`,G=o.div`
  ${n};
  position: relative;
  ${c};
  > ${_} {
    ${l};
  }
`,H=a.keyframes`
  0% {
    transform: perspective(120px) rotateX(0deg) rotateY(0deg);
  }
  50% {
    transform: perspective(120px) rotateX(-180.1deg) rotateY(0deg);
  }
  100% {
    transform: perspective(120px) rotateX(-180deg) rotateY(-179.9deg);
  }
`,L=o.div`
  ${n};
  ${l};
  ${c};
  animation: ${H} ${e=>e.speed}s infinite ease-in-out;
`;exports.ChasingDots=({size:e=40,color:a="#333",className:o})=>t.createElement(y,{className:o,size:e,color:a},t.createElement(u,null),t.createElement(u,{second:!0})),exports.Circle=({size:e=40,color:a="#333",className:o})=>{const s=$(12).map(e=>{const a=r(30*(e+1),1),o=r(1.2/12*e-1.2,1);return t.createElement(N,{key:o,delay:o,transform:a})});return t.createElement(h,{className:o,size:e,color:a},s)},exports.CubeGrid=({size:e=40,color:a="#333",className:o})=>{const s=[.5,.75,1,.25,.5,.75,0,.25,.5].map(e=>r(.4*e,2)).map((e,a)=>t.createElement(P,{key:a,delay:e}));return t.createElement(j,{className:o,size:e,color:a},s)},exports.DoubleBounce=({size:e=40,color:a="#333",className:o})=>t.createElement(G,{className:o,size:e,color:a},t.createElement(_,null),t.createElement(_,{isDelay:!0})),exports.FadingCircle=({size:e=40,color:a="#333",className:o})=>{const s=$(12).map(e=>t.createElement(x,{key:e,rotate:r(30*e,1),delay:r(1.2-1.2/12*e,2)}));return t.createElement(b,{className:o,size:e,color:a},s)},exports.FoldingCube=({size:e=40,color:a="#333",className:o})=>{const s=[0,1,3,2].map(e=>t.createElement(Y,{scale:90*e,key:e,delay:r(.3*e,2),duration:2.4}));return t.createElement(C,{className:o,size:e,color:a},s)},exports.Pulse=({size:e=40,color:a="#333",className:o})=>t.createElement(A,{className:o,size:e,color:a}),exports.RotaingPlane=({size:e=40,color:a="#333",className:o})=>t.createElement(L,{className:o,size:e,color:a,speed:1.2}),exports.ThreeBounce=({size:e=40,color:a="#333",className:o})=>t.createElement(k,{className:o,size:e,color:a},t.createElement(z,{delay:-.32}),t.createElement(z,{delay:-.16}),t.createElement(z,{delay:0})),exports.WanderingCubes=({size:e=40,color:a="#333",className:o})=>{const s=r(e/4,2);return t.createElement(F,{speed:1.8,className:o,size:e,color:a},t.createElement(D,{delay:1.8,size:s}),t.createElement(D,{delay:.9,size:s}))},exports.WaveLoading=({size:e=40,color:a="#333",className:o})=>{const s=$(5).map(e=>t.createElement(M,{delay:r(.1*e-1.2,1),key:e}));return t.createElement(W,{speed:1.2,className:o,size:e,color:a},s)};
